drift/core/psmc.py

```diff
@@ -201,6 +206,356 @@ def _work_fisher_bias_m(self, mi):
+                et_q - st_q,
+            )
+
+        et_allm = time.time()
+        print(
+            "Time needed for quadratic estimation on all ranks for all m-chunks ",
+            et_allm - st,
+        )
+        # Once done with all the m's, redistribute qa array over m's
+        self.qa = self.qa.redistribute(axis=0)
+
+        # Make an array for local fisher an bias
+        fisher_loc = np.zeros((self.nbands, self.nbands), dtype=np.float64)
+        bias_loc = np.zeros((self.nbands,), dtype=np.float64)
+
+        # Calculate fisher for each m
+        for ml, mg in self.qa.enumerate(axis=0):
+            fisher_m = np.cov(self.qa[ml])[: self.nbands, : self.nbands]
+
+            if self.noise:
+                bias_m = fisher_m[-1, : self.nbands]
+            else:
+                bias_m = np.mean(self.qa[ml], axis=1)
+
+            # Sum over all local m-modes to get the overall fisher and bias per parallel process
+            fisher_loc += fisher_m.real  # be careful with the real
+            bias_loc += bias_m.real
+
+        self.fisher = mpiutil.allreduce(fisher_loc, op=MPI.SUM)
+        self.bias = mpiutil.allreduce(bias_loc, op=MPI.SUM)
+
+        if mpiutil.rank0:
+            et = time.time()
+            print("======== Ending PS calculation (time=%f) ========" % (et - st))
+
+        self.write_fisher_file()
+
+    def q_estimator(self, mi, x2, y2=None):
+        """Estimate the q-parameters from given data (see paper).
+
+        Parameters
+        ----------
+        mi : integer
+            The m-mode we are calculating for.
+        x2, y2 : np.ndarrays[nfreq, lmax+1, num_realisations]
+            The vector(s) of data we are estimating from in the sky basis,
+            Passing in `y2` different from `x2` is for cross power spectrum calculation.
+
+        Returns
+        -------
+        qa : np.ndarray[numbands]
+            Array of q-parameters. If noise=True then the array is one longer, and the last parameter is the projection against the noise.
+        """
+
+        # If y2 is None set it to x2
+        if y2 is None:
+            y2 = x2
+
+        # if data vector is filled with zeros, return q = 0.0 for this m
+        if np.all(x2 == 0) or np.all(y2 == 0):
+            self.qa[mi, :] = 0.0
+
+        # If one of the data vectors is empty, return q = 0.0 for this m
+        elif x2.shape[0] == 0:
+            self.qa[mi, :] = 0.0
+
+        else:
+            lside = self.telescope.lmax + 1
+            for bi, bg in self.qa.enumerate(axis=1):
+                for li in range(lside):
+                    lxvec = x2[:, li]
+                    lyvec = y2[:, li]
+                    self.qa[mi, bi] += np.sum(
+                        lyvec.conj()
+                        * np.matmul(self.clarray[bi][li].astype(np.complex128), lxvec),
+                        axis=0,
+                    ).astype(np.float64)
+
+    def noise_bias(self, mi, vec1, vec2=None):
```

Comment:
I added this method for the future, in case we ever want to have a routine to calculate what Richard called the noise bias.

Reply:
Scrap that. I don't think we need it.